File: src/main/java/com/kmarzecki/communicator/security/AuthenticationServiceImpl.java
package com.kmarzecki.communicator.security;

import com.kmarzecki.communicator.exception.OperationNotPermittedException;
import com.kmarzecki.communicator.model.Language;
import com.kmarzecki.communicator.model.auth.LoginDto;
import com.kmarzecki.communicator.model.auth.LoginResponse;
import com.kmarzecki.communicator.model.auth.RegisterDto;
import com.kmarzecki.communicator.model.auth.UserResponse;
import com.kmarzecki.communicator.repository.UserRepository;
import lombok.AllArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.security.Principal;

import static com.kmarzecki.communicator.util.InternationalizationUtil.userExists;

@Service
@AllArgsConstructor
public class AuthenticationServiceImpl implements AuthenticationService {
    private final UserRepository userRepository;
    private final JwtTokenService jwtTokenService;
    private final AuthenticationManager authenticationManager;
    private final UserDetailsServiceImpl userService;

    @Override
    public LoginResponse login(LoginDto dto) {
        String username = dto.getUsername();
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(username, dto.getPassword())
        );
        String token = jwtTokenService.createToken(username, userRepository.findByUsername(username).getRoles());
        return new LoginResponse(username, token);
    }

    @Override
    @Transactional
    public void register(RegisterDto dto) {
        if (userService.existsByUsername(dto.getUsername())) {
                throw new OperationNotPermittedException(userExists(dto.getUsername()));
        }
        userService.saveUser(dto);
    }

    @Override
    public UserResponse getMe(Principal principal) {
        return new UserResponse(principal.getName());
    }
}

File: src/main/kotlin/com/kmarzecki/trollo/security/AuthenticationServiceImpl.kt
package com.kmarzecki.trollo.security

import com.kmarzecki.trollo.exception.OperationNotPermittedException
import com.kmarzecki.trollo.model.auth.LoginDto
import com.kmarzecki.trollo.model.auth.LoginResponse
import com.kmarzecki.trollo.model.auth.RegisterDto
import com.kmarzecki.trollo.model.auth.UserResponse
import com.kmarzecki.trollo.repository.UserRepository
import com.kmarzecki.trollo.util.InternationalizationUtil
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.stereotype.Service
import java.security.Principal
import javax.transaction.Transactional

@Service
class AuthenticationServiceImpl(
        val userRepository: UserRepository,
        val jwtTokenService: JwtTokenService,
        val authenticationManager: AuthenticationManager,
        val userService: UserDetailsServiceImpl
) : AuthenticationService {
    override fun login(dto: LoginDto): LoginResponse {
        val username = dto.username
        authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(username, dto.password)
        )
        val token = jwtTokenService.createToken(username, userRepository.findByUsername(username)!!.roles)
        return LoginResponse(username, token)
    }

    @Transactional
    override fun register(registerDto: RegisterDto) {
        if (userService.existsByUsername(registerDto.username)) {
            throw OperationNotPermittedException(customMessage = InternationalizationUtil.userExists(registerDto.username))
        }
        userService.saveUser(registerDto)
    }

    override fun getMe(principal: Principal): UserResponse {
        return UserResponse(principal.name)
    }
}